File: 30-substring-with-concatenation-of-all-words/substring-with-concatenation-of-all-words.py
class Solution(object):
    def findSubstring(self, s, words):
        """
        :type s: str
        :type words: List[str]
        :rtype: List[int]
        """
        if not s or not words:
            return []
        word_count=Counter(words)
        n=len(words)
        m=len(words[0])
        res=[]
        for i in range(m):
            left=i
            curr_count=Counter()
            count=0
            for j in range(i,len(s)-m+1,m):
                word=s[j:j+m]
                if word in word_count:
                    curr_count[word]+=1
                    count+=1
                    
                    while curr_count[word]>word_count[word]:
                        left_word=s[left:left+m]
                        curr_count[left_word]-=1
                        count-=1
                        left+=m
                    if count==n:
                        res.append(left)
                else:
                    curr_count.clear()
                    count=0
                    left=j+m
        return res



        
      


        

File: 30-substring-with-concatenation-of-all-words/substring-with-concatenation-of-all-words.cpp
class Solution {
public:
    vector<int> findSubstring(string s, vector<string>& words) {
        map<string,int> mp,mp1;
        for(string s:words){
            mp[s]++;
        }
        int n=s.size();
        int m=words[0].size();
        int No_word=words.size();
        vector<int> ans;
        
        
        for(int i=0;i<m;i++){
            mp1.clear();
            int start=i;
            int count=0;
            for(int j=i;j+m<=n;j+=m){
                string st=s.substr(j,m);
                if(mp.count(st)){
                    mp1[st]++;
                    count++;
                    while(mp1[st]>mp[st]){
                        string p=s.substr(start,m);
                        mp1[p]--;
                        start+=m;
                        count--;
                    }
                    if(count==No_word){
                        ans.push_back(start);

                        string p = s.substr(start, m);
                        mp1[p]--;
                        start += m;
                        count--;
                    }
                }else{
                    count=0;
                    start=j+m;
                    mp1.clear();
                }
                
            }
        }
        
        return ans;

    }
};